### ISSUE TITLE
JSONB enters 'null' value into the database for null record

### ISSUE THREAD

**Issue by horalstvo** [jOOQ/jOOQ#9607]
### Expected behavior and actual behavior:
Actual: When one enters null value into a JSONB column like:
```
.set(Tables.MY_TABLE.METADATA, JSONB.valueOf(entry.getMetadata())) 
```
It enters `"null"` into the database instead of `null`.

Expected:
Null is entered into the database when the value passed into `JSONB.valueOf` is null. 

The behavior is also a bit confusing since `JSONB` is just a simple wrapper so one cannot see why would it not handle null well.

Also one would expect that one gets back on the record the wrapper and getting `data()` would be null safe but if there is actually `null` in the database it returns null and one needs to do: 
```
.metadata(record.getMetadata() != null ? record.getMetadata().data() : null)
```

### Steps to reproduce the problem:
https://github.com/horalstvo/jOOQ-mcve
### Versions:

- jOOQ: 3.12.3
- Java: 11
- Database (include vendor): PostgreSQL 10.6
- OS: Ubuntu
- JDBC Driver (include name if inofficial driver): org.postgresql.postgres:42.2.8

**Comment by lukaseder** [comment]
In PostgreSQL (and in other dialects), these three are not the same thing:

```sql
SELECT null, 'null'::jsonb, '"null"'::jsonb
```

The first is an actual SQL `NULL` value. The second is the JSONB `NULL` scalar value. The third one is a JSONB scalar value of type string that happens to look like `null`.

I can see how you're thinking that a `null` string representation of the `JSONB` type would produce a SQL `null` value. But this would be quite orthogonal to everything else in jOOQ. There is not a single other data type where a non-null Java value produces a `NULL` SQL value. You could achieve this using your own custom `Binding`, but it would be quite counter intuitive.

> Also one would expect that one gets back on the record the wrapper and getting data() would be null safe but if there is actually null in the database it returns null and one needs to do:

The same counter intuitive thing would be to return a non-null value in Java, when the corresponding database value is `NULL`. There is not a single other data type in jOOQ that exposes this behaviour. You would start getting used to retrieving "never null" values from jOOQ results, and then suddenly, your `Integer` value is null again.

Three things can be argued here:

1. The status quo should clearly be better documented. I've created a separate issue for this: https://github.com/jOOQ/jOOQ/issues/9612
2. We might argue that a NPE could be thrown when you pass `null` to `JSONB.valueOf()`. I'm not in favour of this. The current behaviour matches that of `String.valueOf()`.
3. We might argue instead that `JSONB.valueOf(null)` would turn `data` to `"null"` explicitly. I will investigate this.

**Comment by horalstvo** [comment]
Sorry for late reply. Thanks for a thorough explanation, documenting the behavior and looking into this. I see what you mean and I agree the semantics should be coherent across jOOQ.

I'm not clear on the JSON semantics of "null". Playing with `jq` it seems rather happy parsing `"null"`. I initially thought that `"null"` is not a valid JSON... but well :) But then maybe actually storing `"null"` in the database is the right thing to do and the problem is on interpreting the value when we retrieve it. As at some point the database value translates into a domain object that can either be `null` reference or an object instance. Then parsing it into the Java object should work with a `"null"` string and return a null reference in that case. Would that make sense? (So the change will be on us and on how we parse the string that came out of the database (from the JSONB column).

So in short it is not clear to me if what we do now is the best way to do this with jOOQ. So on setting:
```
jooq.insertInto(...)
     .set(Tables.MY_TABLE.METADATA, (model.getMetadata() != null) ? JSONB.valueOf(model.getMetadata()) : null)

```
And on retrieval:
```
.metadata(record.getMetadata() != null ? record.getMetadata().data() : null)
```
Maybe we should actually just let it set `"null"` and deal with it when working with the field later.

**Comment by lukaseder** [comment]
> As at some point the database value translates into a domain object that can either be `null` reference or an object instance. Then parsing it into the Java object should work with a `"null"` string and return a null reference in that case. Would that make sense? (So the change will be on us and on how we parse the string that came out of the database (from the JSONB column).

If you want that behaviour, then yes, it will be on you. Again, an actual `NULL` value and a `"null"` json scalar value are not the same thing, regardless of the language, including Java. If you're using `Gson`, there's a difference between having:

```java
JsonElement a = null;
JsonElement b = JsonNull.INSTANCE;
```

They're not the same thing. As a library, we have to be able to represent both values, even if some users prefer not to make a distinction, because those users who like the distinction need a way to represent it.

If you're using JSON binding APIs like Jackson, there are probably ways to achieve what you want. Notice that jOOQ 3.14 will support Jackson if on the classpath: https://github.com/jOOQ/jOOQ/issues/10072

> So in short it is not clear to me if what we do now is the best way to do this with jOOQ.

The question here is whether you need to distinguish between `null` and `"null"`. You probably don't so you could make your columns `NOT NULL`, and maybe use a trigger to change `NULL` to `"null"::jsonb`

**Comment by lukaseder** [comment]
> 3\. We might argue instead that `JSONB.valueOf(null)` would turn `data` to `"null"` explicitly. I will investigate this.

This is a viable solution to the confusing status quo. There's no reason for `JSON::data` or `JSONB::data` to be nullable. Comparing this to my Gson example, in jOOQ, we could have:

```java
JSON a = null
JSON b = JSON.valueOf(null); // Or JSON.valueOf("null"), same thing
```

You can already use `JSON.valueOf(null)` with jOOQ and PostgreSQL, behind the scenes, a `'null'::json` value is created because of some pgjdbc or server side conversion.

It would make sense to reflect this also in `JSON::data`.

I've created a new issue to describe this more clearly: https://github.com/jOOQ/jOOQ/issues/10343. This issue here will be closed as a duplicate of https://github.com/jOOQ/jOOQ/issues/10343 and the previously created documentation issue: https://github.com/jOOQ/jOOQ/issues/9612
